### ISSUE TITLE
Limit range for scale is not used if autoscale is disabled

### ISSUE THREAD

**Issue by espdev** [matplotlib/matplotlib#14515]
### Bug report

**Bug summary**

If we set limits for axis, it disables axis autoscale and limit range for scale will not be checked.
For example, we can decrease zoom and turn on log-scale on plot interactively. If min limit for an axis is less than or equal to zero, we get incorrect plot.

It seems because this code will not be executed:

https://github.com/matplotlib/matplotlib/blob/fac4ad5c016c08ad9efb79fb31550755d9709e86/lib/matplotlib/axes/_base.py#L2447-L2454

https://github.com/matplotlib/matplotlib/blob/fac4ad5c016c08ad9efb79fb31550755d9709e86/lib/matplotlib/axes/_base.py#L2484

and `Axis.limit_range_for_scale` is not used.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(-2., 10., 100)
y = np.sin(x)

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)
ax.plot(x, y, 'o-')

xmin, xmax = ax.get_xlim()

# it disables autoscale (`xmin < 0`)
ax.set_xlim(left=xmin) 
assert not ax.get_autoscalex_on()

print('lin xlim', ax.get_xlim())
print('lin limit range', ax.xaxis.limit_range_for_scale(*ax.get_xlim()))

ax.set_xscale('log')

print('log xlim', ax.get_xlim())
print('log limit range', ax.xaxis.limit_range_for_scale(*ax.get_xlim()))

plt.show()
```

**Actual outcome**

 ![actual](https://user-images.githubusercontent.com/1299189/59178394-89b8e900-8b67-11e9-9570-3cddcafbe070.png)

**Expected outcome**

![expected](https://user-images.githubusercontent.com/1299189/59178363-7b6acd00-8b67-11e9-9c2f-aad06fd7ff01.png)

**Matplotlib version**
3.1

**Comment by timhoffm** [comment]
*TL;DR: I'm not clear what the exact behavior should be here.*

Technically, `set_xlim(left=xmin)` deactivates autoscaling. The question is: Should `set_scale('log')` override this user-defined setting?

- For a fixed script, I would tend to "no". The user explicitly chose a limit and we're not going to override that. For a script one would reasonably set the scale before plotting data.
- The argument may change when considering interactive use where we allow to switch the scale afterwards.

Apart from the conceptual question, there would be some implementation details to consider:

- Should `set_scale('log')` deactivate autoscaling or set a new fixed limit?
  New fixed limit sounds more sane, but OTOH `set_scale('log'); set_scale('linear')` would stick to the log limit.
- What if both limits were fixed negative?
